File: iOS/old/精通iOS开发/8/Cells/Cells/NameAndColorCell.m
//
//  NameAndColorCell.m
//  Cells
//
//  Created by andyron on 16/2/16.
//  Copyright © 2016年 andyron. All rights reserved.
//

#import "NameAndColorCell.h"

// 类扩展
@interface NameAndColorCell ()

@property (strong, nonatomic) IBOutlet UILabel *nameLabel;
@property (strong, nonatomic) IBOutlet UILabel *colorLabel;

@end

@implementation NameAndColorCell
//- (id)initWithStyle:(UITableViewCellStyle)style reuseIdentifier:(NSString *)reuseIdentifier
//{
//    self = [super initWithStyle:style reuseIdentifier:reuseIdentifier];
//    
//    if (self) {
//        // 创建四个UILabel
//        
//        CGRect nameLabelRect = CGRectMake(0, 5, 70, 15);
//        UILabel *nameMarker = [[UILabel alloc] initWithFrame:nameLabelRect];
//        nameMarker.textAlignment = NSTextAlignmentRight;
//        nameMarker.text = @"Name:";
//        nameMarker.font = [UIFont boldSystemFontOfSize:12];
//        [self.contentView addSubview:nameMarker];
//        
//        CGRect colorLabelRect = CGRectMake(0, 26, 70, 15);
//        UILabel *colorMarker = [[UILabel alloc] initWithFrame:colorLabelRect];
//        colorMarker.textAlignment = NSTextAlignmentRight;
//        colorMarker.text = @"Color:";
//        colorMarker.font = [UIFont boldSystemFontOfSize:12];
//        [self.contentView addSubview:colorMarker];
//        
//        CGRect nameValueRect = CGRectMake(80, 5, 200, 15);
//        _nameLabel = [[UILabel alloc] initWithFrame:nameValueRect];
//        [self.contentView addSubview:_nameLabel];
//        
//        CGRect colorValueRect = CGRectMake(80, 25, 200, 15);
//        _colorLabel = [[UILabel alloc] initWithFrame:colorValueRect];
//        [self.contentView addSubview:_colorLabel];
//        
//    }
//    return self;
//}

- (void)setName:(NSString *)n
{
    if (![n isEqualToString:_name]) {
        _name = [n copy];
        self.nameLabel.text = _name;
    }
}
- (void)setColor:(NSString *)c
{
    if (![c isEqualToString:_color]) {
        _color = [c copy];
        self.colorLabel.text = _color;
    }
}
@end

File: iOS/Beginning iPhone Development with Swift 2, 2nd Edition/8/Table Cells/Table Cells/NameAndColorCell.swift
//
//  NameAndColorCell.swift
//  Table Cells
//
//  Created by andyron on 16/7/10.
//  Copyright © 2016年 andyron. All rights reserved.
//

import UIKit

class NameAndColorCell: UITableViewCell {
    
    var name: String = "" {
        didSet {
            if name != oldValue {
                nameLabel?.text = name
            }
        }
    }
    
    /**
     nameLabel.text = name
     
     fatal error: unexpectedly found nil while unwrapping an Optional value
     (lldb)
     
     
     */
    
    var color: String = "" {
        didSet {
            if color != oldValue {
                colorLabel?.text = color
            }
        }
    }
    var nameLabel: UILabel!
    var colorLabel: UILabel!

    override func awakeFromNib() {
        super.awakeFromNib()
        // Initialization code
    }

    override func setSelected(selected: Bool, animated: Bool) {
        super.setSelected(selected, animated: animated)

        // Configure the view for the selected state
    }
    
    override init(style: UITableViewCellStyle, reuseIdentifier: String?) {
        super.init(style: style, reuseIdentifier: reuseIdentifier)
        
        let nameLabelRect = CGRectMake(0, 5, 70, 15)
        let nameMaker = UILabel(frame: nameLabelRect)
        nameMaker.textAlignment = NSTextAlignment.Right
        nameMaker.text = "Name:"
        nameMaker.font = UIFont.boldSystemFontOfSize(12)
        contentView.addSubview(nameMaker)
        
        let colorLabelRect = CGRectMake(0, 26, 70, 15)
        let colorMaker = UILabel(frame: colorLabelRect)
        colorMaker.textAlignment = NSTextAlignment.Right
        colorMaker.text = "Color:"
        colorMaker.font = UIFont.boldSystemFontOfSize(12)
        contentView.addSubview(colorMaker)
        
        let nameValueRect = CGRectMake(80, 5, 200, 15)
        let nameValue = UILabel(frame: nameValueRect)
        contentView.addSubview(nameValue)
        
        let colorValueRect = CGRectMake(80, 26, 200, 15)
        let colorValue = UILabel(frame: colorValueRect)
        contentView.addSubview(colorValue)
        
    }
    //?? 执行的原因
    required init(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }

}
